### PR TITLE
feat: add multiselect to listbox and combobox

### LINKED ISSUE #1103: Combobox Component (Multiselect)
Follow general ARIA practices guidelines: https://www.w3.org/WAI/ARIA/apg/patterns/combobox/

This should include guidelines outlined on #1088 with the addition of the following

Multiselect is configurable via a prop:

```jsx
<Combobox label="Combobox" multiselect>
  <ComboboxItem value="...">
    Item
    Longer Description
  </ComboboxItem>
</Combobox>
```

When multiple items are selected, each item selected should be displayed via a removable tag:

combobox with multiple tags displayed

When the selected items overflows the max-width of the contained area, they should naturally wrap and provide space for the input

combobox with multiple tags that wrap within the text field

## Expected Behaviors

Backspace (with no text entered) should highlight the last selected "pill". Backspace again (or <kbd>delete</kbd>) should delete that selection. When any pill has active focus, backspace or <kbd>delete</kbd> should delete that selection.

Each pill should not be a tabstop but navigated with <kbd>left</kbd> / <kbd>right</kbd> arrow keys. We should support arrow key navigation when the inner text field does not have any text.

When a selection is made to select an active selection, the combobox inner text field should be reset to blank.

When disabled, do not show the "x" for any of the pills and clicking on a pill should not remove it from the selections.

## Form Fields

A single select combobox typically displays form field values as a single hidden input, e.g. `<input type="hidden" name={name} value={value} />`. When multiple values are selected, the hidden input field should be individual fields mapping to the same name for each value, example:

```tsx
{selectedValues.forEach(value => (
  <input type="hidden" name={name} value={value} />
))}
```

### PR DESCRIPTION
**Pull request by chornonoh-vova**
closes #1103

### CODE REVIEW

File: docs/pages/components/Listbox.mdx [hunk 1]
```diff
@@ -127,6 +131,32 @@ Uncontrolled listboxes will automatically set `aria-selected="true"` for the sel
   </Listbox>
 </>
 ```
+### Multiselect
+
+```jsx example
+function MultiselectListboxExample() {
+  const [value, setValue] = useState(['Two']);
+  const handleSelectionChange = ({ value, previousValue }) => {
+      console.log(value, previousValue);
```
Comment (scurker) [inline docs/pages/components/Listbox.mdx:140]: Nit:

```suggestion
```

File: packages/react/scripts/buildIconTypes.js
Comment (scurker) [inline packages/react/scripts/buildIconTypes.js:1]: Do you know why this file changed?

File: packages/react/src/components/Listbox/Listbox.tsx [hunk 2]
```diff
@@ -16,34 +16,43 @@ import useSharedRef from '../../utils/useSharedRef';
...
-  }) => void;
   onActiveChange?: (option: ListboxOption) => void;
-}
+  onSelectionChange?: <T extends HTMLElement = HTMLElement>(selection: {
+    target: T | undefined;
+    previousValue: ListboxValue | ListboxValue[];
+    value: ListboxValue | ListboxValue[];
+  }) => void;
+} & (
+    | {
```
Comment (scurker) [inline packages/react/src/components/Listbox/Listbox.tsx:30]: I'm not sure the discriminated union buys a whole lot here, although if we did want a discriminated union I would make this more readable by keeping the existing type and unioning the types below where needed (example):

```tsx
interfact ListboxProps  = ....
type MultiselectListBoxProps = Omit<ListboxProps, 'value', 'defaultValue', 'multiselect'> & {
  value?: ListboxValue | ListboxValue[]
  defaultValue?: ListboxValue | ListboxValue[]
  multiselect: true
}
```

Part of the reasoning behind allowing array and non-array values is that it allows the component to be more flexible when specifying the value. When referenced, we would use `ListboxProps | MultiselectListboxProps`.

File: packages/react/src/components/Listbox/Listbox.tsx [hunk 3]
```diff
@@ -16,34 +16,43 @@ import useSharedRef from '../../utils/useSharedRef';
...
-    value
-  }: {
-    target: T;
-    previousValue: ListboxValue;
-    value: ListboxValue;
-  }) => void;
   onActiveChange?: (option: ListboxOption) => void;
-}
+  onSelectionChange?: <T extends HTMLElement = HTMLElement>(selection: {
+    target: T | undefined;
```
Comment (scurker) [inline packages/react/src/components/Listbox/Listbox.tsx:25]: Why `| undefined`?

File: packages/react/src/components/Listbox/Listbox.tsx [hunk 4]
```diff
@@ -65,9 +75,9 @@ const Listbox = forwardRef<HTMLElement, ListboxProps>(
     const [activeOption, setActiveOption] = useState<ListboxOption | null>(
       null
     );
-    const [selectedOption, setSelectedOption] = useState<ListboxOption | null>(
-      null
-    );
+    const [selectedOption, setSelectedOption] = useState<
```
Comment (scurker) [inline packages/react/src/components/Listbox/Listbox.tsx:78]: Since we potentially want to handle array values, I would change this to be `selectedOptions` and `setSelectedOptions` because the existing naming strategy doesn't make sense for both single/multi select.

File: packages/react/src/components/Listbox/Listbox.tsx [hunk 5]
```diff
@@ -179,13 +239,13 @@ const Listbox = forwardRef<HTMLElement, ListboxProps>(
...
           );
-          // istanbul ignore else
           if (firstOption) {
             setActiveOption(firstOption);
           }
-          // istanbul ignore else
         } else if (event.target === listboxRef.current) {
-          setActiveOption(selectedOption);
+          setActiveOption(
+            Array.isArray(selectedOption) ? selectedOption[0] : selectedOption
```
Comment (scurker) [inline packages/react/src/components/Listbox/Listbox.tsx:247]: Ideally, this would be keying off the prop and not the value type.

```tsx
!multiselect ? selectedOption[0] : selectedOption
```

File: packages/react/src/components/Listbox/ListboxOption.tsx [hunk 6]
```diff
@@ -68,14 +77,11 @@ const ListboxOption = forwardRef<HTMLElement, ListboxOptionProps>(
           return [...options, option];
         }
 
-        for (const opt of options) {
+        for (let i = 0; i < options.length; i++) {
```
Comment (scurker) [inline packages/react/src/components/Listbox/ListboxOption.tsx:80]: Why the change here?

File: packages/styles/combobox.css [hunk 7]
```diff
@@ -111,8 +111,28 @@
...
 
+.Combobox .TextFieldWrapper {
+  flex-wrap: wrap;
+}
+
 .Combobox .TextFieldWrapper input[type='text'] {
   margin-right: 1.25rem;
+  flex-grow: 1;
+  width: unset;
+  min-width: var(--input-min-width);
```
Comment (scurker) [inline packages/styles/combobox.css:122]: Why was `min-width` added here? This creates an accessibility issue at lower viewport widths.

File: packages/styles/combobox.css [hunk 8]
```diff
@@ -111,8 +111,28 @@
...
+}
+
+.Combobox__selected-tag {
+  margin-block: var(--space-smallest);
```
Comment (scurker) [inline packages/styles/combobox.css:126]: I'm not sure this is correct as the vertical spacing between the selected items doesn't appear to match the designs.

Actual:
image

Proposed Design:
image

The down carat is also not positioned correctly in the top right.

File: packages/styles/combobox.css [hunk 9]
```diff
@@ -111,8 +111,28 @@
...
+}
+
+.Combobox__selected-tag:first-of-type {
+  margin-left: var(--space-smallest);
+}
+
+.Combobox__selected-tag .TagButton__icon svg {
+  width: 16px;
+  height: 16px;
```
Comment (scurker) [inline packages/styles/combobox.css:135]: I think instead of defining the size here, we may be able to just ensure we're using `--icon-size: 16px` scoped to the right css selector.

File: packages/react/src/components/Combobox/Combobox.tsx [hunk 10]
```diff
@@ -46,15 +47,38 @@ interface ComboboxProps
...
+    value: ComboboxValue | ComboboxValue[];
+    previousValue: ComboboxValue | ComboboxValue[];
   }) => void;
   onActiveChange?: (option: ListboxOption) => void;
   renderNoResults?: (() => JSX.Element) | React.ReactElement;
   portal?: React.RefObject<HTMLElement> | HTMLElement;
   inputRef?: React.Ref<HTMLInputElement>;
-}
+} & (
+    | {
```
Comment (scurker) [inline packages/react/src/components/Combobox/Combobox.tsx:59]: Same comment as above regarding Listbox and its discriminated unions.

File: packages/react/src/components/Combobox/Combobox.tsx [hunk 11]
```diff
@@ -30,14 +32,13 @@ interface ComboboxOption {
...
-  > {
+type ComboboxProps = Omit<
+  React.InputHTMLAttributes<HTMLInputElement>,
+  'value' | 'defaultValue'
+> & {
   label: ContentNode;
   options?: ComboboxOption[];
-  value?: ComboboxValue;
-  defaultValue?: ComboboxValue;
+  inputValue?: string;
```
Comment (scurker) [inline packages/react/src/components/Combobox/Combobox.tsx:41]: I apologize because this wasn't well defined, but I'm not sure we should be changing `value` to `inputValue` here. Existing users of the component are expecting to use `value` for non-multiselect components so we need to maintain that compatibility in order to not be a breaking change.

What we may need is a mechanism for converting a value string into a multiselect value. Feel free to update the RFC as we need to reach a consensus to ensure the changes we make here are compatible and make sense.

**Review by scurker — CHANGES_REQUESTED** [review]
There's a lot of great work here, but I'm realizing looking at your PR that we missed defining some of the more refined behaviors of how we expect users to interact with this component. I left some comments inline of some of the things I noticed, but we may need a few rounds to ensure we land on the right spot here.

File: packages/react/scripts/buildIconTypes.js
Comment (chornonoh-vova) [reply packages/react/scripts/buildIconTypes.js:1]: This file was changed after running `yarn fmt`, actually a couple of other files in `packages/react/__test__` also changed, but I didn't include them

File: packages/react/src/components/Listbox/Listbox.tsx (hunk 2 above)
Comment (chornonoh-vova) [reply packages/react/src/components/Listbox/Listbox.tsx:30]: By doing this I'm only allowing `value: string[]` when `multiselect = true` and `value: string` when `multiselect = false`. I don't really understand why we should allow value to be both string and string[] when multiselect is true, behavior will just be broken in my opinion. But if that's required, I'll change that.

File: packages/react/src/components/Listbox/Listbox.tsx (hunk 4 above)
Comment (chornonoh-vova) [reply packages/react/src/components/Listbox/Listbox.tsx:78]: But in the single select case name `selectedOptions` doesn't make sense either 😅

File: packages/react/src/components/Combobox/Combobox.tsx (hunk 11 above)
Comment (chornonoh-vova) [reply packages/react/src/components/Combobox/Combobox.tsx:41]: I had no idea how to convert value into multiselect value, therefore added this prop. It's just not compatible - `onChange` emits `string` and `onSelectionChange` emits `string[]` values. One of the possible solutions - albeit radical one - would be only left `value` and `onSelectionChange` props, and remove `onChange`. Although I don't know if that's plausible for users of the component.

File: packages/react/src/components/Listbox/ListboxOption.tsx (hunk 6 above)
Comment (chornonoh-vova) [reply packages/react/src/components/Listbox/ListboxOption.tsx:80]: It's just easier, previously we had line `const index = options.indexOf(opt);` inside of the loop, so why not have this index already when iterating?

File: packages/react/src/components/Combobox/Combobox.tsx (hunk 11 above)
Comment (longmatthewh) [reply packages/react/src/components/Combobox/Combobox.tsx:42]: @scurker what would you think about creating a new `MultiSelectCombobox` component? I don't totally love the idea (feels a little lazy), but I also don't hate it. Thoughts?

File: packages/react/scripts/buildIconTypes.js
Comment (scurker) [reply packages/react/scripts/buildIconTypes.js:1]: We really want to keep the changes focused to what needs to change for this component and isolate changes that are unrelated as a separate PR/issue.

File: packages/react/src/components/Listbox/Listbox.tsx (hunk 2 above)
Comment (scurker) [reply packages/react/src/components/Listbox/Listbox.tsx]: Part of the reasoning would be to allow for more graceful inputs since we can't necessarily enforce that everyone is using typescript. Someone using `<Combobox value="foo" multiselect />` at a minimum should not cause an error although we could potentially log a warning if this wasn't the desired interface.

File: packages/react/src/components/Listbox/Listbox.tsx (hunk 4 above)
Comment (scurker) [reply packages/react/src/components/Listbox/Listbox.tsx]: My thinking here was if the internal state is always an array, then it's always "selected options of 1" but I'm open to discussion here.

File: packages/react/src/components/Combobox/Combobox.tsx (hunk 11 above)
Comment (scurker) [reply packages/react/src/components/Combobox/Combobox.tsx:42]: > @scurker what would you think about creating a new MultiSelectCombobox component? I don't totally love the idea (feels a little lazy), but I also don't hate it. Thoughts?

There's a lot of overlap between the single/multiselect components, so that introduces an additional challenge of keeping those components in sync if we split them up with future updates. It's also a higher lift for authors since swapping between a single/multiselect component is potentially as small as changing the property with the current proposal.

Comment (scurker) [reply packages/react/src/components/Combobox/Combobox.tsx:42]: > I had no idea how to convert value into multiselect value, therefore added this prop. It's just not compatible - onChange emits string and onSelectionChange emits string[] values.

That's where the discriminated union could potentially help, example below.

```tsx
type ComboboxProps = {
  ...
  onChange: (value: string, ...) => {} 
}

type MultiselectComboboxProps extends Omit<ComboboxProps, 'onChange'> = {
  ...
  onChange: (value: string[]) => {}
}
```

File: packages/react/src/components/Combobox/ComboboxOption.tsx [hunk 12]
```diff
@@ -113,10 +123,23 @@ const ComboboxOption = forwardRef<HTMLLIElement, ComboboxOptionProps>(
...
+        if ((selectedValue as ComboboxValue[]).includes(comboboxValue)) {
+          setFormValue((prev) => {
+            const formValueToSet =
+              typeof formValue === 'undefined' ? comboboxValue : formValue;
+
+            if (!prev?.includes(formValueToSet)) {
+              return prev ? prev + ',' + formValueToSet : formValueToSet;
+            }
+          });
+        }
```
Comment (scurker) [inline packages/react/src/components/Combobox/ComboboxOption.tsx:142]: This is worth discussing, I'm not sure what the best approach is (and it needs to be addressed as part of the RFC). There's 2 distinct approaches we could take:

- Comma delimited form values
- Individual form values

It appears you've taken the first approach, but what if the value itself includes a comma? It can introduce challenges since the server will need to be aware of the specific format to parse by and be able to translate comma delimited values correctly.

Alternatively, each value itself could be a set of hidden fields that map to the same name:

```html
<input type="hidden" name={name} value={value1} />
<input type="hidden" name={name} value={value2} />
<input type="hidden" name={name} value={value3} />
```

I feel like this approach is more scalable and avoids having to directly parse the submitted values. Thoughts?

File: packages/react/src/components/Listbox/ListboxOption.tsx (hunk 6 above)
Comment (scurker) [reply packages/react/src/components/Listbox/ListboxOption.tsx:80]: I'm partially nitpicking here because it's not clear why the change was made and isn't conventionally how iterations have been performed elsewhere. I understand this can be subjective, but given there's not a problem that's being identified it would be better to leave this untouched unless we're fixing a specific issue.

File: packages/react/src/components/Combobox/ComboboxOption.tsx (hunk 12 above)
Comment (chornonoh-vova) [reply packages/react/src/components/Combobox/ComboboxOption.tsx:142]: I really like this approach, thanks for suggestion

File: packages/styles/combobox.css [hunk 13]
```diff
@@ -111,8 +111,24 @@
   position: relative;
 }
 
+.Combobox .TextFieldWrapper {
+  min-height: 32px;
+  flex-wrap: wrap;
+  gap: var(--space-half);
+  padding-inline: var(--space-smallest);
+  padding-block: var(--space-half);
```
Comment (scurker) [inline packages/styles/combobox.css:119]: Is there a reason why these styles were added? This appears to change the size of all Comboboxes. Comboboxes should match the existing size of other comboboxes (which is currently 35px).

File: packages/styles/combobox.css [hunk 14]
```diff
@@ -111,8 +111,24 @@
...
+
 .Combobox .TextFieldWrapper input[type='text'] {
   margin-right: 1.25rem;
+  flex-grow: 1;
+  width: unset;
+  padding: unset;
+}
+
+.Combobox__selected-tag {
+  margin-left: 0 !important;
```
Comment (scurker) [inline packages/styles/combobox.css:130]: Why `!important`? This should be limited to extremely specific use cases (such as utility classes) where the desired behavior is to take on that style without respect of specificity. There are likely better alternatives that we need to consider before using `!important`.
